File: 백준/Silver/1699. 제곱수의 합/제곱수의 합.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    
    int n;
    cin >> n;

    vector<int> dp(n + 1, 100000);
    dp[0] = 0;
    
    vector<int> squares;
    for (int i = 1; i * i <= n; i++) {
        squares.push_back(i * i);
    }

    for (int i = 1; i <= n; i++) {
        for (int square : squares) {
            if (square > i)
                break;
            dp[i] = min(dp[i], dp[i - square] + 1);
        }
    }

    cout << dp[n] << "\n";
    return 0;
}

File: 백준/Silver/1699. 제곱수의 합/제곱수의 합.js
const fs = require("fs");

const n = Number(fs.readFileSync(0, "utf8").trim());

const dp = [];
for (let i = 0; i < n + 1; i++)
    dp[i] = i;

const squares = [];
for (let i = 1; i * i <= n; i++) {
    squares.push(i * i);
}

for (let i = 2; i <= n; i++) {
    for (const square of squares) {
        if (square > i)
            break;
        dp[i] = Math.min(dp[i], dp[i - square] + 1);
    }
}

console.log(dp[n]);
